llms/mlx_lm/models/llama.py

```diff
@@ -30,12 +32,128 @@ def __post_init__(self):
+            if not any(key in self.rope_scaling for key in ["type", "rope_type"]):
+                raise ValueError(
+                    f"rope_scaling must contain either 'type' or 'rope_type'"
+                )
+            if not "factor" in self.rope_scaling:
+                raise ValueError(f"rope_scaling must contain 'factor'")
+            rope_type = self.rope_scaling.get("type") or self.rope_scaling.get(
+                "rope_type"
+            )
+            if rope_type not in ["linear", "dynamic", "llama3"]:
+                raise ValueError(
+                    "rope_scaling 'type' currently only supports 'linear', 'dynamic' or 'llama3'"
+                )
+
+
+class DynamicNTKScalingRoPE(nn.Module):
+    """Implements the rotary positional encoding with Dynamic NTK scaling and Llama 3 RoPE."""
+
+    def __init__(
+        self,
+        dims: int,
+        max_position_embeddings: int = 2048,
+        traditional: bool = False,
+        base: float = 10000,
+        scale: float = 1.0,
+        rope_type: str = "default",
+        rope_scaling: dict = None,
+    ):
+        super().__init__()
+        self.dims = dims
+        self.max_position_embeddings = max_position_embeddings
+        self.traditional = traditional
+        self.original_base = base
+        self.scale = scale
+        self.rope_type = rope_type
+        self.rope_scaling = rope_scaling
+
+    def compute_base_freq(self):
+        if self.rope_type == "llama3":
+            return self.compute_llama3_base_freq()
+        return self.original_base
+
+    def compute_llama3_base_freq(self):
+        factor = self.rope_scaling["factor"]
+        low_freq_factor = self.rope_scaling.get("low_freq_factor", 1.0)
+        high_freq_factor = self.rope_scaling.get("high_freq_factor", 4.0)
+        old_context_len = self.rope_scaling.get(
+            "original_max_position_embeddings", 8192
+        )
 
-            if self.rope_scaling["type"] != "linear":
-                raise ValueError("rope_scaling 'type' currently only supports 'linear'")
+        low_freq_wavelen = old_context_len / low_freq_factor
+        high_freq_wavelen = old_context_len / high_freq_factor
+
+        freqs = self.original_base ** (
+            mx.arange(0, self.dims, 2).astype(mx.float32) / self.dims
+        )
+        new_base_freqs = []
+
+        for freq in freqs:
+            wavelen = 2 * math.pi * freq
+            if wavelen < high_freq_wavelen:
+                new_base_freqs.append(freq)
+            elif wavelen > low_freq_wavelen:
+                new_base_freqs.append(freq * factor)
+            else:
+                assert low_freq_wavelen != high_freq_wavelen
+                smooth = (wavelen - high_freq_wavelen) / (
+                    low_freq_wavelen - high_freq_wavelen
+                )
+                new_base_freqs.append(freq * ((1 - smooth) * factor + smooth))
+
+        return mx.array(new_base_freqs).mean().item()
```

Comment:
It's not so great to have a graph evaluation at each call to RopE (e.g. `item`). That will be slow. This might not matter too much since it only applies for very long sequence lengths. But might be worth leaving a TODO here to fix this if we enable RoPE to take `base` as an `mx.array`.

Reply:
Added a TODO for now.

Reply:
That makes sense! Thanks for clarifying @awni.

I noticed the slow speed but didn't have enough time to debug yesterday as I was heading out.

Could be a good MLX lesson Twitter post.

Reply:
> Could be a good MLX lesson Twitter post.

Definitely ;)

Reply:
Please tag me :)